File: frontend/src/app/services/turma.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Turma } from '../core/models';

@Injectable({ providedIn: 'root' })
export class TurmaService {
  private readonly http = inject(HttpClient);
  private readonly base = '/api/turmas';

  listar(disciplinaId?: number): Observable<Turma[]> {
    let params = new HttpParams();
    if (disciplinaId != null) {
      params = params.set('disciplinaId', disciplinaId);
    }
    return this.http.get<Turma[]>(this.base, { params });
  }

  criar(payload: {
    disciplinaId: number;
    codigo: string;
    periodo: string;
    vagasTotais: number;
  }): Observable<Turma> {
    return this.http.post<Turma>(this.base, payload);
  }

  atualizar(
    id: number,
    payload: { codigo: string; periodo: string; vagasTotais: number }
  ): Observable<Turma> {
    return this.http.put<Turma>(`${this.base}/${id}`, payload);
  }

  abrir(id: number): Observable<Turma> {
    return this.http.post<Turma>(`${this.base}/${id}/abrir`, {});
  }

  fechar(id: number): Observable<Turma> {
    return this.http.post<Turma>(`${this.base}/${id}/fechar`, {});
  }

  excluir(id: number): Observable<void> {
    return this.http.delete<void>(`${this.base}/${id}`);
  }
}

File: backend/src/main/java/br/com/techne/matriculas/turma/service/TurmaService.java
package br.com.techne.matriculas.turma.service;

import br.com.techne.matriculas.disciplina.domain.Disciplina;
import br.com.techne.matriculas.disciplina.domain.DisciplinaRepository;
import br.com.techne.matriculas.shared.exception.RecursoNaoEncontradoException;
import br.com.techne.matriculas.turma.domain.Turma;
import br.com.techne.matriculas.turma.domain.TurmaRepository;
import java.util.List;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class TurmaService {

  private final TurmaRepository turmaRepository;
  private final DisciplinaRepository disciplinaRepository;

  public TurmaService(TurmaRepository turmaRepository, DisciplinaRepository disciplinaRepository) {
    this.turmaRepository = turmaRepository;
    this.disciplinaRepository = disciplinaRepository;
  }

  @Transactional(readOnly = true)
  public List<Turma> listar() {
    return turmaRepository.findAllWithDisciplina();
  }

  @Transactional(readOnly = true)
  public List<Turma> listarPorDisciplina(Long disciplinaId) {
    if (!disciplinaRepository.existsById(disciplinaId)) {
      throw new RecursoNaoEncontradoException("Disciplina", disciplinaId);
    }
    return turmaRepository.findByDisciplinaIdWithDisciplina(disciplinaId);
  }

  @Transactional(readOnly = true)
  public Turma buscar(Long id) {
    return turmaRepository.findByIdWithDisciplina(id)
        .orElseThrow(() -> new RecursoNaoEncontradoException("Turma", id));
  }

  @Transactional
  public Turma criar(Long disciplinaId, String codigo, String periodo, Integer vagasTotais) {
    Disciplina disciplina = disciplinaRepository.findByIdWithCurso(disciplinaId)
        .orElseThrow(() -> new RecursoNaoEncontradoException("Disciplina", disciplinaId));
    Turma salva = turmaRepository.save(new Turma(disciplina, codigo.trim(), periodo.trim(), vagasTotais));
    return buscar(salva.getId());
  }

  @Transactional
  public Turma atualizar(Long id, String codigo, String periodo, Integer vagasTotais) {
    Turma turma = buscar(id);
    turma.atualizar(codigo.trim(), periodo.trim(), vagasTotais);
    turmaRepository.save(turma);
    return buscar(id);
  }

  @Transactional
  public Turma abrir(Long id) {
    Turma turma = buscar(id);
    turma.abrir();
    turmaRepository.save(turma);
    return buscar(id);
  }

  @Transactional
  public Turma fechar(Long id) {
    Turma turma = buscar(id);
    turma.fechar();
    turmaRepository.save(turma);
    return buscar(id);
  }

  @Transactional
  public void excluir(Long id) {
    turmaRepository.delete(buscar(id));
  }
}
